// chart-component/src/app/app.component.ts
import {Component, OnInit, ViewChild} from '@angular/core';
import {ILineData} from "./chart-module/multi-line-chart/multi-line-chart.component";
import {forEach} from "lodash";
import {Http} from "@angular/http";

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss']
})
export class AppComponent implements OnInit {

  public lineDataObjects: Array<ILineData> = [
    {
      color: "blue",
      name: "Scope",
      shown: true,
      data: null
    },
    {
      color: "red",
      name: "Dark",
      shown: true,
      data: null
    }
  ];

  public thumbnailDataObjects: Array<ILineData> = [
    {
      color: "red",
      name: "Dark",
      shown: true,
      data: null
    }
  ];

  public width: number= 500;

  @ViewChild('darkChart')
  public darkChart: any;
  @ViewChild('thumbnailChart')
  public thumbnailChart: any;

  constructor(protected http: Http) {
  }

  public ngOnInit() {

  }

  public lineToggled (index: number) {
    this.lineDataObjects[index].shown = !this.lineDataObjects[index].shown;

    let array = [];
    forEach(this.lineDataObjects, (temp) => {
      array.push(temp);
    });

    this.lineDataObjects = array;
  }
}
