// ngx-mat-multi-sort/esm2015/lib/mat-multi-sort-header/mat-multi-sort-header.component.js
import { Component, Input, ChangeDetectorRef, Optional, Inject, HostListener, HostBinding, ViewEncapsulation, ChangeDetectionStrategy, ElementRef } from '@angular/core';
import { matSortAnimations, MatSortHeader, MatSortHeaderIntl } from '@angular/material/sort';
import { MatMultiSort } from '../mat-multi-sort.directive';
import { FocusMonitor } from '@angular/cdk/a11y';
export class MatMultiSortHeaderComponent extends MatSortHeader {
    constructor(_intl, changeDetectorRef, _sort, _columnDef, _focusMonitor, _elementRef) {
        super(_intl, changeDetectorRef, _sort, _columnDef, _focusMonitor, _elementRef);
        this._intl = _intl;
        this._sort = _sort;
        this._columnDef = _columnDef;
        this.start = 'asc';
    }
    __setIndicatorHintVisible(visible) {
        super._setIndicatorHintVisible(visible);
    }
    _handleClick() {
        this._sort.direction = this.getSortDirection();
        super._handleClick();
    }
    _isSorted() {
        return this._sort.actives.findIndex(activeId => activeId === this.id) > -1;
    }
    _sortId() {
        return this._sort.actives.findIndex(activeId => activeId === this.id) + 1;
    }
    _updateArrowDirection() {
        this._arrowDirection = this.getSortDirection();
    }
    _getAriaSortAttribute() {
        if (!this._isSorted()) {
            return null;
        }
        return this.getSortDirection() === 'asc' ? 'ascending' : 'descending';
    }
    _renderArrow() {
        return !this._isDisabled() || this._isSorted();
    }
    getSortDirection() {
        const i = this._sort.actives.findIndex(activeIds => activeIds === this.id);
        const direction = this._sort.directions[i];
        return this._isSorted() ? direction : (this.start || this._sort.start);
    }
}
MatMultiSortHeaderComponent.decorators = [
    { type: Component, args: [{
                selector: '[mat-multi-sort-header]',
                exportAs: 'matMultiSortHeader',
                template: "<div class=\"mat-sort-header-container\"\r\n     [class.mat-sort-header-sorted]=\"_isSorted()\"\r\n     [class.mat-sort-header-position-before]=\"arrowPosition == 'before'\">\r\n  <button class=\"mat-sort-header-button\" type=\"button\"\r\n          [attr.disabled]=\"_isDisabled() || null\"\r\n          [attr.aria-label]=\"_intl.sortButtonLabel(id)\"\r\n          (focus)=\"_setIndicatorHintVisible(true)\"\r\n          (blur)=\"_setIndicatorHintVisible(false)\">\r\n    <ng-content></ng-content>\r\n  </button>\r\n\r\n  <!-- Disable animations while a current animation is running -->\r\n  <div class=\"mat-sort-header-arrow\"\r\n       *ngIf=\"_renderArrow()\"\r\n       [@arrowOpacity]=\"_getArrowViewState()\"\r\n       [@arrowPosition]=\"_getArrowViewState()\"\r\n       [@allowChildren]=\"_getArrowDirectionState()\"\r\n       (@arrowPosition.start)=\"_disableViewStateAnimation = true\"\r\n       (@arrowPosition.done)=\"_disableViewStateAnimation = false\">\r\n    <div class=\"mat-sort-header-stem\"></div>\r\n    <div class=\"mat-sort-header-indicator\" [@indicator]=\"_getArrowDirectionState()\">\r\n      <div class=\"mat-sort-header-pointer-left\" [@leftPointer]=\"_getArrowDirectionState()\"></div>\r\n      <div class=\"mat-sort-header-pointer-right\" [@rightPointer]=\"_getArrowDirectionState()\"></div>\r\n      <div class=\"mat-sort-header-pointer-middle\"></div>\r\n    </div>\r\n  </div>\r\n  <div *ngIf=\"_isSorted()\">{{_sortId()}}</div>\r\n</div>\r\n",
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                animations: [
                    matSortAnimations.indicator,
                    matSortAnimations.leftPointer,
                    matSortAnimations.rightPointer,
                    matSortAnimations.arrowOpacity,
                    matSortAnimations.arrowPosition,
                    matSortAnimations.allowChildren
                ],
                styles: [".mat-sort-header-container{align-items:center;cursor:pointer;display:flex}.mat-sort-header-disabled .mat-sort-header-container{cursor:default}.mat-sort-header-position-before{flex-direction:row-reverse}.mat-sort-header-button{align-items:center;background:0 0;border:none;color:currentColor;cursor:inherit;display:flex;font:inherit;outline:0;padding:0}.mat-sort-header-button::-moz-focus-inner{border:0}.mat-sort-header-arrow{display:flex;height:12px;min-width:12px;opacity:0;position:relative;width:12px}.mat-sort-header-arrow,[dir=rtl] .mat-sort-header-position-before .mat-sort-header-arrow{margin:0 0 0 6px}.mat-sort-header-position-before .mat-sort-header-arrow,[dir=rtl] .mat-sort-header-arrow{margin:0 6px 0 0}.mat-sort-header-stem{align-items:center;background:currentColor;display:flex;height:10px;margin:auto;width:2px}.mat-sort-header-indicator{align-items:center;display:flex;height:2px;left:0;position:absolute;top:0;width:100%}.mat-sort-header-pointer-middle{background:currentColor;height:2px;margin:auto;transform:rotate(45deg);width:2px}.mat-sort-header-pointer-left,.mat-sort-header-pointer-right{background:currentColor;height:2px;position:absolute;top:0;width:6px}.mat-sort-header-pointer-left{left:0;transform-origin:right}.mat-sort-header-pointer-right{right:0;transform-origin:left}"]
            },] }
];
MatMultiSortHeaderComponent.ctorParameters = () => [
    { type: MatSortHeaderIntl },
    { type: ChangeDetectorRef },
    { type: MatMultiSort, decorators: [{ type: Optional }] },
    { type: undefined, decorators: [{ type: Inject, args: ['C2_SORT_HEADER_COLUMN_DEF',] }, { type: Optional }] },
    { type: FocusMonitor },
    { type: ElementRef }
];
MatMultiSortHeaderComponent.propDecorators = {
    id: [{ type: Input, args: ['mat-multi-sort-header',] }],
    __setIndicatorHintVisible: [{ type: HostListener, args: ['mouseenter', ['true'],] }, { type: HostListener, args: ['longpress', ['true'],] }, { type: HostListener, args: ['mouseleave', ['false'],] }],
    _getAriaSortAttribute: [{ type: HostBinding, args: ['attr.aria-sort',] }]
};
